runtime/lua/vim/_editor.lua

```diff
@@ -535,6 +535,55 @@ function vim.region(bufnr, pos1, pos2, regtype, inclusive)
   return region
 end
 
+--- Gets the content of the visual selection.
+---
+--- The result is either a string or, if {list} is `true`, a list of strings.
+--- If not in any |visual-mode|, `nil` is returned.
+---
+--- @param list boolean|nil
+---        Return a list of strings instead of a string.  See |getreg()|.
+---        Defaults to `false`.
+--- @param append_empty boolean|nil
+---        Append an empty string to the result when in |linewise-visual| mode and {list} is `true`.
+---        This will preserve the trailing newline of the selection when the result is concatenated with `"\n"`.
+---        Defaults to `false`.
+--- @return string|table
+function vim.get_visual_selection(list, append_empty)
```

Comment:
Same comment that I made in https://github.com/neovim/neovim/pull/13896 and https://github.com/neovim/neovim/pull/21115#issuecomment-1322153511 

In particular #21115 has test coverage and re-uses low-level C internals which is likely a preferred starting point. 

Adding a new public function for _this_ PR can be avoided. I've linked to https://github.com/justinmk/config/blob/6e591bad58db8200e8eeebe20982ce3d1fc09c91/.config/nvim/init.vim#L794-L816 several times, why not use it? It's less code.

Reply:
Thanks for looking at this.

I first tried using `vim.region` but it was challenging when the last character of a selection was multibyte.  IIRC `vim.region` returns the byte index of the last character selected, but I need to know how many bytes are in that character (to add to the `end_col` param in `nvim_buf_get_text`).

I called `nvim_buf_get_lines` to get the whole line, but I'd still need to find how many bytes are in the last character of the selection.
Given the whole line, I tried using `strgetchar`, `nr2char`, `byteidx`, and `string.sub` in various ways, but it got complicated.

I eventually gave up and used the yank approach because it just works.

Some of my pain points when using the `vim.region`, `nvim_buf_get_lines`, `nvim_buf_get_text` approach:
* `nvim_buf_get_text` does't have a `strict_indexing` option like `nvim_buf_get_lines` has.
* Lua 5.1 isn't Unicode-aware
* The APIs take/return 0-index values, but Lua arrays, row numbers, and column numbers start at 1.  (I know I should "git gud", but I was always second-guessing if I needed to add or subtract 1)

Reply:
Those are valid problems, but let's solve them instead of adding a new api.

> nvim_buf_get_text doesn't have a strict_indexing option like nvim_buf_get_lines has. 

Could add a `opts.strict_indexing` flag to nvim_buf_get_text.

> Lua 5.1 isn't Unicode-aware

There is no such thing really ([see also](https://github.com/neovim/neovim/issues/68#issuecomment-170268097)). But `strgetchar`, `strchars`, `strwidth` are available from Lua.